File: 55. Jump Game/main.py
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        size = len(nums)
        canReach = [False] * size

        canReach[0] = True

        for i in range(size):
            if not canReach[i]:
                return False

            for j in range(i, min(i + nums[i] + 1, size)):
                canReach[j] = True

        return True


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        maxReach = 0

        for i in range(0, len(nums)):
            if i > maxReach:
                return False

            maxReach = max(maxReach, i + nums[i])

        return True


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        size = len(nums)
        start = size - 2
        target = size - 1

        while start >= 0:
            if start + nums[start] >= target:
                target = start

            start -= 1

        return target == 0


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        size = len(nums)
        target = size - 1

        for start in range(size - 2, -1, -1):
            if start + nums[start] >= target:
                target = start

        return target == 0

File: 55. Jump Game/main.rs
struct Solution;

impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {
        let size = nums.len();
        let mut can_reach = vec![false; size];

        can_reach[0] = true;
        for i in 0..size {
            if !can_reach[i] {
                return false;
            }

            for j in i..size.min(nums[i] as usize + i + 1) {
                can_reach[j] = true;
            }
        }

        true
    }
}

impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {
        let size = nums.len();
        let mut max_reach = 0;

        for i in 0..size {
            if i > max_reach {
                return false;
            }

            max_reach = max_reach.max(i + nums[i] as usize);
        }

        true
    }
}

impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {
        let size = nums.len();
        let mut target = size as i32 - 1;
        let mut start = size as i32 - 2;

        while start >= 0 {
            if start + nums[start as usize] >= target {
                target = start;
            }
            start -= 1;
        }

        target == 0
    }
}
